Internal: Add useSearch [ED-20930]

MuliElementor opened the pull request:
add search functional

## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

<!--start_gitstream_placeholder-->
### ✨ PR Description
```
Purpose: Add search functionality to Components tab with debounced input, filtering, and proper UI feedback for empty states.
Main changes:
- Implemented SearchProvider context with debounced search functionality using useSearchState hook
- Added ComponentSearch component with search input field and icon
- Created filtered components list that updates based on search term
- Added empty search results state with clear search option
```

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using. **[We'd love your feedback!](mailto:product@linearb.io)** 🚀</sub>
<!--end_gitstream_placeholder-->

MuliElementor committed:
add useSearch hook
add search functional

elementorbot committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
add tests

MuliElementor committed:
update tests

MuliElementor committed:
Update component search and item components to use 'small' font size for icons; adjust tests for component visibility

elementorbot committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
Refactor: Update styles and types in ComponentItem and settings API

- Adjusted padding and border properties in the ComponentItem for improved layout.
- Changed icon size from 'small' to 'tiny' for consistency.
- Updated the type returned by getSettings from Settings to Homepage for better type accuracy.

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component

# Conflicts:
#	packages/packages/core/editor-components/src/components/components-item.tsx

elementorbot committed:
Merge branch 'main' into ED-20930-search-component

elementorbot committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
Merge branch 'main' of https://github.com/elementor/elementor into ED-20930-search-component

MuliElementor committed:
update

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
update

MuliElementor committed:
Add EmptySearchResult component for improved search feedback

Introduce an EmptySearchResult component to handle cases where no components match the current search query. This improves the user experience by providing clear feedback and allowing users to clear their search and try again. Additionally, updated `useFilteredComponents` to include `searchValue`.

MuliElementor committed:
update

MuliElementor committed:
update

IshayMaya commented on packages/packages/libs/utils/src/search-provider/search-provider.tsx:
```diff
@@ -0,0 +1,55 @@
+	children,
+	localStorageKey,
+}: {
+	children: React.ReactNode;
+	localStorageKey?: string;
+} ) => {
+	const getInitialSearchValue = () => {
+		if ( localStorageKey ) {
+			const storedValue = localStorage.getItem( localStorageKey );
+			if ( storedValue ) {
+				localStorage.removeItem( localStorageKey );
+				return storedValue;
+			}
+		}
+		return '';
+	};
+
+	const onClearSearch = () => {
+		handleChange( '' );
+	};
+
+	const { debouncedValue, inputValue, handleChange } = useDebounceState( {
+		delay: 300,
+		initialValue: getInitialSearchValue(),
+	} );
+
+	return (
+		<SearchContext.Provider value={ { debouncedValue, inputValue, handleChange, onClearSearch } }>
+			{ children }
+		</SearchContext.Provider>
+	);
+};
+
+export const useSearch = () => {
+	const context = useContext( SearchContext );
+	if ( ! context ) {
+		throw new Error( 'useSearch must be used within a SearchContextProvider' );
+	}
+	return context;
+};
```
i don't think we should have util that is binded to a provider, maybe just export the hook from here and then anyone can implement it however they want to

```suggestion

function useSearchState( { localStorageKey }: { localStorageKey: string } ) {
	const getInitialSearchValue = () => {
		if ( localStorageKey ) {
			const storedValue = localStorage.getItem( localStorageKey );
			if ( storedValue ) {
				localStorage.removeItem( localStorageKey );
				return storedValue;
			}
		}
		return '';
	};

	const { debouncedValue, inputValue, handleChange } = useDebounceState( {
		delay: 300,
		initialValue: getInitialSearchValue(),
	} );

	return {
		debouncedValue,
		inputValue,
		handleChange,
	};
}

```

IshayMaya commented on packages/packages/core/editor-components/src/components/components-list.tsx:
```diff
@@ -1,25 +1,30 @@
 		return <LoadingComponents />;
 	}
 
+	if ( searchValue.length > 0 && ( ! components || components.length === 0 ) ) {
+		return <EmptySearchResult />;
+	}
+
 	if ( ! components || components.length === 0 ) {
 		return <EmptyState />;
 	}
```
```suggestion
	const isEmpty = ! components || components.length === 0;

	if ( isEmpty ) {
		if ( searchValue.length > 0 ) {
			return <EmptySearchResult searchValue={ searchValue } handleChange={ handleChange } />;
		}

		return <EmptyState />;
	}
```

IshayMaya commented on packages/packages/core/editor-components/src/components/components.tsx:
```diff
@@ -1,12 +1,17 @@
+import { ComponentSearch } from './component-search';
 import { ComponentsList } from './components-list';
 
 export const Components = () => {
```
maybe we dont need provider yet, we can always add it if needed
```
export const Components = () => {
	const { debouncedValue, inputValue, handleChange } = useSearchState( {
		localStorageKey: 'elementor-components-search',
	} );

	return (
		<ThemeProvider>
			<ComponentSearch inputValue={ inputValue } handleChange={ handleChange } />
			<ComponentsList searchValue={ debouncedValue } handleChange={ handleChange } />
		</ThemeProvider>
	);
};
```

IshayMaya commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -35,6 +41,13 @@ const mockComponents: Component[] = [
 describe( 'ComponentsList', () => {
 	beforeEach( () => {
 		jest.clearAllMocks();
+		// Default mock for useSearch - no search active
```
please remove unnecessary comments, and add the "Arrange/Act/Assert" to the tests

IshayMaya commented on packages/packages/core/editor-components/src/components/component-search.tsx:
```diff
@@ -0,0 +1,31 @@
+					InputProps={ {
+						startAdornment: (
+							<InputAdornment position="start">
+								<SearchIcon fontSize={ 'small' } />
```
the text field is tiny and icon is small? check if it fits the figma

IshayMaya commented on packages/packages/libs/utils/src/__tests__/search-provider.test.tsx:
```diff
@@ -0,0 +1,164 @@
+		expect( result.current.debouncedValue ).toBe( '' );
+	} );
+
+	it( 'should handle localStorage when no stored value exists', () => {
```
not sure it need to be tested

IshayMaya commented on packages/packages/libs/utils/src/__tests__/search-provider.test.tsx:
```diff
@@ -0,0 +1,164 @@
+
+import { SearchProvider, useSearch } from '../search-provider/search-provider';
+
+// Mock localStorage
```
here as well

IshayMaya commented on packages/packages/libs/utils/src/__tests__/search-provider.test.tsx:
```diff
@@ -0,0 +1,164 @@
+		expect( result.current.debouncedValue ).toBe( 'debounced search' );
+	} );
+
+	it( 'should cancel previous debounce and keep the latest value', () => {
```
`useDebounceState` is already tested

MuliElementor committed:
Add components for enhanced component management

Introduce several new components for the editor's component tab, including `ComponentSearch`, `ComponentsList`, `ComponentItem`, and `LoadingComponents`. These components facilitate searching, displaying, and managing components within the editor. Additionally, a new `SearchProvider` is implemented to manage search state effectively, replacing the previous search context. This update enhances user experience by providing a structured and interactive way to handle components.

MuliElementor committed:
Refactor component structure and enhance search functionality

This commit reorganizes the component structure by moving components into a dedicated `components-tab` directory. It introduces a new `SearchProvider` to manage search state, replacing the previous implementation. The `ComponentSearch`, `ComponentsList`, and `ComponentItem` components have been updated to utilize the new search context, improving the overall search experience. Additionally, various tests have been added to ensure the functionality of the new search state management.

MuliElementor committed:
Enhance component tests and search functionality

This commit updates the test suite for the `ComponentsList` and `ComponentItem` components, ensuring comprehensive coverage of loading states, empty states, and search functionality. The `useSearch` hook is integrated into the tests to simulate various search scenarios, improving the accuracy of component behavior under different conditions. Additionally, the structure of the tests is refined for better readability and maintainability.

MuliElementor committed:
update

MuliElementor committed:
Update useSearchState to allow optional localStorageKey parameter

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
lint

IshayMaya commented on packages/packages/libs/utils/src/__tests__/use-search-state.test.tsx:
```diff
@@ -0,0 +1,25 @@
+import { useSearchState } from '../index';
+
+describe( 'useSearchState', () => {
+	it( 'should return the expected interface', () => {
```
maybe test the actual functionality? typescript already cover the interfaces

IshayMaya commented on packages/packages/core/editor-components/src/components/components-tab/search-provider.tsx:
```diff
@@ -0,0 +1,38 @@
+} ) => {
+	const { debouncedValue, handleChange, inputValue } = useSearchState( { localStorageKey } );
+
+	const onClearSearch = () => {
```
```suggestion
	const clearSearch = () => {
```
since it the actual action and not a callback

IshayMaya commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -115,24 +132,23 @@ describe( 'ComponentsList', () => {
 		} );
 
 		it( 'should render component name', () => {
+			// Arrange.
```
```suggestion
```
try to go through the tests, the comments are missing or partials in some

IshayMaya commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -18,8 +18,17 @@ jest.mock( '../../utils/get-container-for-new-element' );
+	SearchProvider: ( { children }: { children: React.ReactNode } ) => children,
 	useSearch: jest.fn(),
 } ) );
+jest.mock( '@elementor/editor-ui', () => ( {
```
why? we have `renderWithTheme` for that

IshayMaya commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -18,8 +18,17 @@ jest.mock( '../../utils/get-container-for-new-element' );
 	...jest.requireActual( '@elementor/utils' ),
+	useSearchState: jest.fn(),
+} ) );
+jest.mock( '../components-tab/search-provider', () => ( {
```
since its part of the tabs functionality now, i think it should be part of the tests and not a mock (wrap the component with the actual provider)

MuliElementor committed:
fix comment

MuliElementor committed:
Refactor and enhance component tests for improved search functionality

This commit refines the test suite for the `ComponentsList`, `ComponentItem`, and related components, ensuring comprehensive coverage of loading states, empty states, and search functionality. The integration of the `SearchProvider` and updates to the `useSearch` hook enhance the accuracy of component behavior under various conditions. Additionally, the structure of the tests has been improved for better readability and maintainability.

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
update

MuliElementor committed:
pull

MuliElementor committed:
Enhance tests for useSearchState hook with improved localStorage handling

This commit updates the test suite for the `useSearchState` hook, adding comprehensive tests for various scenarios including initialization with empty values, handling stored values, and debounced updates. It also introduces error handling for localStorage failures and ensures that the hook behaves correctly without a localStorage key. The structure of the tests has been refined for better clarity and maintainability.

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component

MuliElementor committed:
Merge branch 'main' into ED-20930-search-component
